File: Leetcode/654/654.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */


// Even though you could use Sparse Table for maximal range-queries quickly,
// the precomputation still cost O(nlogn)
class Solution {
public:
    TreeNode* createNode(vector<int>& nums, int i, int j) {
        if (i > j) {
            return nullptr;
        }
        auto iter = max_element(nums.begin()+i,nums.begin()+j+1);
        int pos = (int)(iter - nums.begin());
        TreeNode* node = new TreeNode(*iter);
        // divide
        TreeNode* left = createNode(nums, i, pos-1);
        TreeNode* right = createNode(nums, pos+1, j);
        // conquer
        node->left = left;
        node->right = right;
        return node;
    }
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return createNode(nums, 0, (int)nums.size()-1);
    }
};

File: Leetcode/654/654.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxElement(nums []int, l int, r int) (int, int) {
	if l > r {
		panic("l should be not larger than r!")
	}
	ans, pos := math.MinInt32, l
	for i := l; i <= r; i++ {
		if nums[i] > ans {
			ans = nums[i]
			pos = i
		}
	}
	return ans, pos
}
func solve(nums []int, l int, r int) *TreeNode {
	if l > r {
		return nil
	}
	val, pos := maxElement(nums, l, r)
	node := new(TreeNode)
	node.Val = val
	node.Left = solve(nums, l, pos-1)
	node.Right = solve(nums, pos+1, r)
	return node
}
func constructMaximumBinaryTree(nums []int) *TreeNode {
	return solve(nums, 0, len(nums)-1)
}